### PR TITLE
feat(model): add `recompileSchema()` function to models to allow applying schema changes after compiling

### LINKED ISSUE #14296: Virtual declared after model creation, doesn't work
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

8.1.1

### Node.js version

20.11.0

### MongoDB server version

6.0.12

### Typescript version (if applicable)

_No response_

### Description

If I define a virtual after model declaration, the virtual only works calling```toJSON()``` to document.

For now I've to use ```applyGetters()``` to the virtualType to get the value.

### Steps to Reproduce

```js
import db from "mongoose";
//await db.connect("mongodb://127.0.0.1:27017/test");
const Schema = new db.Schema({
  field: String
});
Schema.virtual("virtual1").get(function () {
  return this.field + " virtual1";
});
Schema.virtual("virtual2").get(function () {
  return this.virtualAfter + " virtual2";
});
Schema.virtual("virtual3").get(function () {
  return this.constructor.schema.virtualpath("virtual4").applyGetters(undefined, this) + " virtual3";
});
const Model = db.model("model", Schema);
Model.schema.virtual("virtual4").get(function () {
  return this.field + " virtual4";
});

const doc = new Model({field: "test"});
console.log("doc", doc);
console.log("doc JSON", doc.toJSON({virtuals: true}));
console.log("doc.virtual1", doc.virtual1);
console.log("doc.virtual2", doc.virtual2);
console.log("doc.virtual3", doc.virtual3);
console.log("doc.virtual4", doc.virtual4);
console.log("doc.virtual1 applyGetters", Schema.virtualpath("virtual1").applyGetters(undefined, doc));
console.log("doc.virtual2 applyGetters", Schema.virtualpath("virtual2").applyGetters(undefined, doc));
console.log("doc.virtual3 applyGetters", Schema.virtualpath("virtual3").applyGetters(undefined, doc));
console.log("doc.virtual4 applyGetters", Schema.virtualpath("virtual4").applyGetters(undefined, doc));
```

### Expected Behavior

If ```doc.toJSON()``` works with a virtual defined after model, I should be able to get the virtual directly.

In the docs the only restriction is to avoid hook definitions after model.

### PR DESCRIPTION
**Pull request by vkarpov15**
Fix #14296

**Summary**

Making changes to schemas after compiling a model is a common feature request. We can support that by adding an extra model function `recompileSchema()` that applies schema changes after model compilation.

**Examples**

### CODE REVIEW

File: test/model.test.js [hunk 1]
```diff
@@ -7302,6 +7302,19 @@ describe('Model', function() {
       /First argument to `Model` constructor must be an object/
     );
   });
+
```
Comment (AbdelrahmanHafez) [inline test/model.test.js:7305]: Suggestion: would add another test case verifying that the schema isn't modified without calling `recompileSchema`. Unless that's implicitly tested somewhere else.

**Review by AbdelrahmanHafez — APPROVED** [review]
One minor suggestion, otherwise LGTM. Thanks! 👍

File: lib/model.js [hunk 2]
```diff
@@ -4861,6 +4861,34 @@ Model.__subclass = function subclass(conn, schema, collection) {
   return Model;
 };
 
+/**
+ * Apply changes made to this model's schema after this model was compiled.
+ * By default, adding virtuals and other properties to a schema after the model is compiled does nothing.
+ * Call this function to apply virtuals and properties that were added later.
+ *
+ * #### Example:
+ *     const schema = new mongoose.Schema({ field: String });
```
Comment (hasezoey) [inline lib/model.js:4870]:
```suggestion
 * #### Example:
 *
 *     const schema = new mongoose.Schema({ field: String });
```

File: lib/model.js [hunk 3]
```diff
@@ -4861,6 +4861,34 @@ Model.__subclass = function subclass(conn, schema, collection) {
...
+ *     TestModel.schema.virtual('myVirtual').get(function() {
+ *       return this.field + ' from myVirtual';
+ *     });
+ *     const doc = new TestModel({ field: 'Hello' });
+ *     doc.myVirtual; // undefined
+ *
+ *     TestModel.recompileSchema();
+ *     doc.myVirtual; // 'Hello from myVirtual'
+ *
+ * @return {Model}
```
Comment (hasezoey) [inline lib/model.js:4881]: this function does not seem to return anything though?

**Review by hasezoey — CHANGES_REQUESTED** [review]
Looks good, though the jsdoc return type is wrong and addition to typescript types is missing

File: test/model.test.js (hunk 1 above)
Comment (vkarpov15) [reply test/model.test.js:7305]: There's a part in the "supports recompiling model with new schema additions" test case that asserts `myVirtual()` isn't applied before `recompileSchema()` is called.

File: types/models.d.ts [hunk 4]
```diff
@@ -711,6 +711,8 @@ declare module 'mongoose' {
       options?: (mongodb.ReplaceOptions & MongooseQueryOptions<TRawDocType>) | null
     ): QueryWithHelpers<UpdateWriteOpResult, ResultDoc, TQueryHelpers, TRawDocType, 'replaceOne'>;
 
+    recompileSchema(): void;
```
Comment (hasezoey) [inline types/models.d.ts:714]: should likely still have some tsdoc

**Review by hasezoey — APPROVED** [review]

**Comment by vkarpov15** [comment]
Merging into 8.2

**Pull request merged**